cedar-policy-core/src/ast/policy.rs

```diff
@@ -943,91 +992,221 @@ pub struct TemplateBody {
     non_scope_constraints: Arc<Expr>,
 }
 
+#[derive(Clone, Hash, Eq, PartialEq, Debug)]
+/// Policy datatype. This is used for both templates (in which case it contains
+/// slots) and static policies (in which case it contains zero slots).
+pub enum TemplateBody {
+    /// Represents a valid template body
+    TemplateBody(TemplateBodyImpl),
+    #[cfg(feature = "tolerant-ast")]
+    /// Represents a policy that failed to parse
+    TemplateBodyError(PolicyID),
+}
+
+impl Serialize for TemplateBody {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            TemplateBody::TemplateBody(impl_body) => impl_body.serialize(serializer),
+            #[cfg(feature = "tolerant-ast")]
+            TemplateBody::TemplateBodyError(policy_id) => {
+                format!("TemplateBody::TemplateBodyError({policy_id})").serialize(serializer)
+            }
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for TemplateBody {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let impl_body = TemplateBodyImpl::deserialize(deserializer)?;
+        Ok(TemplateBody::TemplateBody(impl_body))
+    }
+}
+
 impl TemplateBody {
     /// Get the `Id` of this policy.
     pub fn id(&self) -> &PolicyID {
-        &self.id
+        match self {
+            TemplateBody::TemplateBody(TemplateBodyImpl { id, .. }) => &id,
+            #[cfg(feature = "tolerant-ast")]
+            TemplateBody::TemplateBodyError(policy_id) => &policy_id,
+        }
     }
 
     /// Get the location of this policy
     pub fn loc(&self) -> Option<&Loc> {
-        self.loc.as_ref()
+        match self {
+            TemplateBody::TemplateBody(TemplateBodyImpl { loc, .. }) => loc.as_ref(),
+            #[cfg(feature = "tolerant-ast")]
+            TemplateBody::TemplateBodyError(_) => None,
```

Comment:
perhaps we could propagate the `Loc` through `TemplateBodyError` so we can actually have a `Loc` associated with an error policy?

Reply:
Works for me! I will add